server/util/rexec/rexec.go

```diff
@@ -0,0 +1,282 @@
+// package rexec provides a collection of utility functions for remote execution
+// clients.
+package rexec
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/buildbuddy-io/buildbuddy/server/environment"
+	"github.com/buildbuddy-io/buildbuddy/server/interfaces"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/cachetools"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/digest"
+	"github.com/buildbuddy-io/buildbuddy/server/util/retry"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"golang.org/x/exp/maps"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/genproto/googleapis/longrunning"
+
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	rspb "github.com/buildbuddy-io/buildbuddy/proto/resource"
+	gstatus "google.golang.org/grpc/status"
+)
+
+// AssembleEnv assembles a list of EnvironmentVariable protos from a list of
+// NAME=VALUE pairs. If the same name is specified more than once, the last one
+// wins. The entries are sorted by name, so that the environment variables are
+// cache-friendly.
+func AssembleEnv(pairs ...string) ([]*repb.Command_EnvironmentVariable, error) {
+	m := map[string]string{}
+	for _, pair := range pairs {
+		parts := strings.SplitN(pair, "=", 2)
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("invalid environment variable %q (expected NAME=VALUE)", pair)
+		}
+		m[parts[0]] = parts[1]
+	}
+	names := maps.Keys(m)
+	sort.Strings(names)
+	out := make([]*repb.Command_EnvironmentVariable, 0, len(m))
+	for _, name := range names {
+		out = append(out, &repb.Command_EnvironmentVariable{
+			Name:  name,
+			Value: m[name],
+		})
+	}
+	return out, nil
+}
+
+// AssemblePlatform assembles a Platform proto from a list of NAME=VALUE pairs.
+// If the same name is specified more than once, the last one wins. The entries
+// are sorted by name, so that the platform is cache-friendly.
+func AssemblePlatform(pairs ...string) (*repb.Platform, error) {
+	m := map[string]string{}
+	for _, s := range pairs {
+		parts := strings.SplitN(s, "=", 2)
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("invalid exec property %q (expected NAME=VALUE)", s)
+		}
+		m[parts[0]] = parts[1]
+	}
```

Comment:
not a huge deal but if you wanted to pull this into a helper function you could (it's the same code here and above?)

func parsePairs(pairs []string) (map[string]string, error) {
	m := map[string]string{}
	for _, s := range pairs {
		parts := strings.SplitN(s, "=", 2)
		if len(parts) != 2 {
			return nil, return status.InvalidInputError("invalid pair %q, expected key=val format")
		}
		m[parts[0]] = parts[1]
	}
    return m, nil
}

Reply:
good call, done